code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,470 @@
+			if(!touch_desc.on_touch(src, src))
+				touch_desc = null //not setting this to null fucks with check, qdel refuses to be helpful another day
+
+	special_desc = initial(desc_holder?.desc) ? "[special_desc] [initial(desc_holder.desc)] material." : "[special_desc] material."
+
+	if(malf)
+		generate_trait_unique(XENOA_MALFS, blacklist_traits) //Malf
+
+	desc_holder = generate_trait_unique(XENOA_MAJORS, blacklist_traits, FALSE) //Major
+	special_desc = initial(desc_holder.desc) ? "[special_desc] The shape is [initial(desc_holder.desc)]." : "[special_desc] The shape is Unknown."
+
+	charge_req = rand(1, 10) * 10
+
+///generate a single trait against a blacklist. Used in larger /obj/item/xenoartifact/proc/generate_traits()
+/obj/item/xenoartifact/proc/generate_trait_unique(var/list/trait_list, var/list/blacklist_traits = list())
+	var/datum/xenoartifact_trait/new_trait //Selection
+	var/list/selection = trait_list //Selectable traits
+	selection -= blacklist_traits
+	if(selection.len < 1)
+		log_game("An almost impossible event has occured. [src] has failed to generate any traits with [trait_list]!")
+		return
+	new_trait = pickweight(selection)
+	blacklist_traits += new_trait //Add chosen trait to blacklist
+	traits += new new_trait
+	new_trait = new new_trait //type converting doesn't work too well here but this should be fine.
+	blacklist_traits += new_trait.blacklist_traits //Cant use initial() to access lists without bork'ing it
+	return new_trait
+	
+///Gets a singular entity, there's a specific traits that handles multiple.
+/obj/item/xenoartifact/proc/get_target_in_proximity(range)
+	for(var/mob/living/M in oview(range, get_turf(src)))
+		. = process_target(M)
+	if(isliving(loc) && !.)
+		. = loc
+	return
+
+///Returns the desired trait and it's values if it's in the artifact's list
+/obj/item/xenoartifact/proc/get_trait(typepath)
+	return (locate(typepath) in traits)
+
+///Used for hand-holding secret technique. Pulling entities swaps them for you in the target list.
+/obj/item/xenoartifact/proc/process_target(atom/target)
+	if(ishuman(target)) //early return if deflect chance
+		var/mob/living/carbon/human/H = target
+		if(H.wear_suit && H.head && isclothing(H.wear_suit) && isclothing(H.head))
+			var/obj/item/clothing/CS = H.wear_suit
+			var/obj/item/clothing/CH = H.head
+			if(((CS.clothing_flags & BLOCK_ARTIFACT)||(CH.clothing_flags & BLOCK_ARTIFACT)) && prob(XENOA_DEFLECT_CHANCE))
+				to_chat(target, "<span class='warning'>The [name] was unable to target you!.</span>")
+				playsound(get_turf(target), 'sound/weapons/deflect.ogg', 35, TRUE) 
+				return
+
+	if(isliving(target)) //handle pulling
+		var/mob/living/M = target
+		. = M?.pulling ? M.pulling : M
+	else
+		. = target
+	RegisterSignal(., COMSIG_PARENT_QDELETING, .proc/on_target_del, .)
+	return
+
+///Hard del handle
+/obj/item/xenoartifact/proc/on_target_del(atom/target)
+	UnregisterSignal(target, COMSIG_PARENT_QDELETING)
+	true_target -= target
+	target = null
```

Comment:
Setting a local variable to null is pointless, the reference will disappear when the variable goes out of scope.